smarts/core/smarts.py

```diff
@@ -312,7 +326,26 @@ def done_agent_ids(dones):
         self._teardown_vehicles(vehicles_to_teardown)
 
     def reset(self, scenario: Scenario):
-        if scenario == self._scenario and self._reset_agents_only:
+        tries = 2
+        first_exception = None
+        for _ in range(tries):
+            try:
+                self._resetting = True
+                return self._reset(scenario)
+            except Exception as e:
+                if not first_exception:
+                    first_exception = e
+            finally:
+                self._resetting = False
+        self._log.error(f"Failed to successfully reset after {tries} times.")
+        raise first_exception
```

Comment:
I am somewhat iffy about this change intended to give a slight leeway in case an edge case occurs that breaks `reset`. Essentially, because we do not return `dones` or `infos` from this `reset` it is not easily possible to communicate to the user that `reset` failed for some reason. 

There are two directions we can take with this: 
- The `reset` has the responsibility to give a working simulation state. (currently going with this)
- The user has to determine if `reset` worked.

Reply:
I think you're right to go with the first one.